src/views/catalog/templatescatalog/hooks/useAvailableDataSourcesAndPVCs.ts

```diff
@@ -1,91 +1,100 @@
 import React from 'react';
 
-import { V1Template } from '@kubevirt-ui/kubevirt-api/console';
+import { PersistentVolumeClaimModel, V1Template } from '@kubevirt-ui/kubevirt-api/console';
+import DataSourceModel from '@kubevirt-ui/kubevirt-api/console/models/DataSourceModel';
 import { V1beta1DataSource } from '@kubevirt-ui/kubevirt-api/containerized-data-importer/models';
+import { IoK8sApiCoreV1PersistentVolumeClaim } from '@kubevirt-ui/kubevirt-api/kubernetes';
 import { BOOT_SOURCE } from '@kubevirt-utils/resources/template';
 import {
-  getDataSource,
-  getPVC,
   getTemplateBootSourceType,
   isDataSourceReady,
 } from '@kubevirt-utils/resources/template/hooks/useVmTemplateSource/utils';
-
-import { assertFulfilled, getSourcePromises } from './utils';
+import {
+  getGroupVersionKindForModel,
+  useK8sWatchResources,
+  WatchK8sResource,
+} from '@openshift-console/dynamic-plugin-sdk';
 
 type UniqueSourceType = {
-  [namespace: string]: Set<string>;
+  [key in string]: WatchK8sResource;
 };
 
 /**
  * Hook that returns the DataSources and PVCs that are available for the templates
  * @param templates - the templates to filter
  * @param templatesLoaded - whether the templates are loaded
- * @returns availablePVCs and availableDatasources, both Sets of strings representing the available sources. `{namespace-name}`
+ * @returns availablePVCs Set of strings representing the available pvcs. `{namespace-name}`, availableDatasources record with `{namespace-name}` as keys and data sources as values.
  */
 export const useAvailableDataSourcesAndPVCs = (
   templates: V1Template[],
   templatesLoaded: boolean,
 ) => {
-  const [availableDatasources, setAvailableDatasources] = React.useState<
-    Record<string, V1beta1DataSource>
-  >({});
-  const [availablePVCs, setAvailablePVCs] = React.useState<Set<string>>();
-  const [loaded, setLoaded] = React.useState(false);
-
   const uniqueSources = React.useMemo(() => {
-    if (templatesLoaded) {
-      const dataSources: UniqueSourceType = {};
-      const pvcs: UniqueSourceType = {};
+    if (!templatesLoaded)
+      return {
+        uniqueDataSources: {},
+        uniquePVCs: {},
+      };
 
-      templates.forEach((t) => {
-        const bootSource = getTemplateBootSourceType(t);
+    const dataSources: UniqueSourceType = {};
+    const pvcs: UniqueSourceType = {};
 
-        if (bootSource.type === BOOT_SOURCE.DATA_SOURCE) {
-          const ds = bootSource?.source?.sourceRef;
-          dataSources[ds?.namespace] = new Set([...(dataSources?.[ds?.namespace] || []), ds?.name]);
-        }
+    templates.forEach((t) => {
+      const bootSource = getTemplateBootSourceType(t);
 
-        if (bootSource.type === BOOT_SOURCE.PVC) {
-          const pvc = bootSource?.source?.pvc;
-          pvcs[pvc?.namespace] = new Set([...(pvcs?.[pvc?.namespace] || []), pvc?.name]);
-        }
-      });
+      if (bootSource.type === BOOT_SOURCE.DATA_SOURCE) {
```

Comment:
Do we have guaranteed that `bootSource` will always have `type`? Not sure if this won't crash, even if it's part of the if-block...

Reply:
there will always be a type. The worst case would be  `NONE`

Reply:
video attached

Reply:
I think u can change to reduce